middleman_protocol/tests/tests.py

```diff
@@ -0,0 +1,453 @@
+                is_frame_separator_expected=use_separator,
+                is_encoding_expected=use_encoding,
+            )
+
+        assertpy.assert_that(str(expected_exception.value)).is_equal_to('Frame signature does not match its content.')
+
+    def test_that_serializing_and_deserializing_message_with_wrong_payload_type_should_raise_exception(self):
+        middleman_message = GolemMessageMiddlemanMessage(Ping(), self.request_id)
+        raw_message = middleman_message.serialize(private_key=CONCENT_PRIVATE_KEY)
+
+        payload_type_position = FRAME_SIGNATURE_BYTES_LENGTH + FRAME_REQUEST_ID_BYTES_LENGTH
+
+        # Sanity check for payload type in this case to be between expected bytes
+        assert raw_message[payload_type_position:payload_type_position + 1] == b'\x00'
+        assert 100 not in PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS
+
+        # Replace bytes with payload length
+        raw_message = (
+            raw_message[:payload_type_position] +
+            bytes(bytearray([100])) +
+            raw_message[payload_type_position + 1:]
+        )
+
+        # Replace message signature
+        new_signature = ecdsa_sign(CONCENT_PRIVATE_KEY, raw_message[FRAME_SIGNATURE_BYTES_LENGTH:])
+        raw_message_with_new_signature = new_signature + raw_message[FRAME_SIGNATURE_BYTES_LENGTH:]
+
+        with pytest.raises(MiddlemanProtocolError) as expected_exception:
+            AbstractMiddlemanMessage.deserialize(raw_message_with_new_signature, CONCENT_PUBLIC_KEY)
+
+        assertpy.assert_that(
+            str(expected_exception.value)
+        ).is_equal_to(
+            'Payload type 100 is not valid or not registered.'
+        )
+
+
+class TestHandleMessageReceiveHelper:
+
+    request_id = 99
+
+    @classmethod
+    def _replace_payload_length_and_payload_to_length_one(cls, raw_message):
+        return (
+            raw_message[:FRAME_PAYLOAD_LENGTH_STARTING_BYTE] +
+            b'\x00\x00\x00\x01' +
+            bytes(bytearray([0]))
+        )
+
+    @pytest.mark.parametrize(('use_separator', 'use_encoding'), [
+        (False, False),
+        (False, True),
+        (True, False),
+        (True, True),
+    ])
+    @pytest.mark.parametrize(('middleman_message_type', 'payload'), [
+        (GolemMessageMiddlemanMessage,            Ping()),
+        (ErrorMiddlemanMessage,                   (111, 'error_message')),
+        (AuthenticationChallengeMiddlemanMessage, b'random_bytes'),
+        (AuthenticationResponseMiddlemanMessage,  b'TODO'),
+    ])
+    def test_that_receiving_any_message_should_be_handled_correctly(
+        self,
+        middleman_message_type,
+        payload,
+        use_separator,
+        use_encoding,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+        raw_message = middleman_message.serialize(
+            private_key=CONCENT_PRIVATE_KEY,
+            add_frame_separator=use_separator,
+        )
+
+        if use_encoding:
```

Comment:
@rwrzesien. this `if ... else` seems unnecessary - you can pass `use_encoding` to ` middleman_message.serialize()`

Reply:
:+1: done